src/lib/stable/canister_methods/query.ts

```diff
@@ -1,52 +1,28 @@
 import { IDL } from '@dfinity/candid';
 
-import { handleUncaughtError } from '../error';
-import { executeAndReplyWithCandidSerde } from '../execute_with_candid_serde';
+import { decoratorArgumentsHandler, DecoratorFunction, MethodType } from '.';
 
 export function query<This, Args extends any[], Return>(
-    paramIdlTypes: IDL.Type[],
+    originalMethod: MethodType<This, Args, Return>,
+    context: ClassMethodDecoratorContext
+): MethodType<This, Args, Return>;
+
+export function query<This, Args extends any[], Return>(
+    paramIdlTypes?: IDL.Type[],
     returnIdlType?: IDL.Type,
     options?: {
         composite?: boolean;
         manual?: boolean;
     }
-) {
-    return (
-        originalMethod: (this: This, ...args: Args) => Return,
-        context: ClassMethodDecoratorContext
-    ): void => {
-        const index = globalThis._azleCanisterMethodsIndex++;
-        const name = context.name as string;
-        const composite = options?.composite ?? false;
-        const indexString = index.toString();
-
-        globalThis._azleMethodMeta.queries?.push({
-            name,
-            index,
-            composite
-        });
+): (
+    originalMethod: MethodType<This, Args, Return>,
+    context: ClassMethodDecoratorContext
+) => MethodType<This, Args, Return>;
 
-        globalThis._azleCanisterMethodIdlTypes[name] = IDL.Func(
-            paramIdlTypes,
-            returnIdlType === undefined ? [] : [returnIdlType],
-            ['query']
-        );
-
-        globalThis._azleCallbacks[indexString] = async (
-            args: Uint8Array
-        ): Promise<void> => {
-            try {
-                await executeAndReplyWithCandidSerde(
-                    'query',
-                    args,
-                    originalMethod.bind(globalThis._azleCanisterClassInstance),
-                    paramIdlTypes,
-                    returnIdlType,
-                    options?.manual ?? false
-                );
-            } catch (error: any) {
-                handleUncaughtError(error);
-            }
-        };
-    };
+export function query<This, Args extends any[], Return>(
```

Comment:
I don't understand why we have this third version, I'm guessing the first is for not invoking and the second is for invoking?

Reply:
I see that this is an implementation handler. which I understand at a high level but I don't see how this implementation works